File: app/src/main/java/com/ndmsystems/coala/layers/arq/SlidingWindow.java
package com.ndmsystems.coala.layers.arq;

import com.ndmsystems.coala.helpers.logging.LogHelper;

/**
 * Created by Владимир on 16.08.2017.
 */

public class SlidingWindow<T> {

    protected int offset;
    protected Object[] values;
    private int size;

    public SlidingWindow(int size, int offset) {
        this.offset = offset;
        values = new Object[size];
    }

    public SlidingWindow(int size) {
        this(size, 0);
    }

    public int getOffset() {
        return offset;
    }

    public int getSize() {
        return values.length;
    }

    public void set(int number, T value) {
        int windowIndex = number - offset;
        if (windowIndex > values.length - 1) {
            LogHelper.e("ARQ: window index out of bounds");
            return;
        } else if (windowIndex < 0)
            return;

        values[windowIndex] = value;
    }

    public T advance() {
        T firstBlock = (T) values[0];

        if (firstBlock == null)
            return null;

        System.arraycopy(values, 1, values, 0, values.length - 1);
        values[values.length - 1] = null;
        offset++;

        return firstBlock;
    }

    public T getValue(int windowIndex) {
        return (T) values[windowIndex];
    }

    public int tail() {
        return offset + values.length - 1;
    }

    public void setSize(int windowSize) {
        Object[] tempValues = new Object[windowSize];
        System.arraycopy(values, 0, tempValues, 0, Math.min(windowSize, size));
        values = tempValues;
        size = windowSize;
    }
}

File: app/src/main/java/com/ndmsystems/coala/layers/arq/SlidingWindow.kt
package com.ndmsystems.coala.layers.arq

import com.ndmsystems.coala.helpers.logging.LogHelper.e
import kotlin.math.min

/**
 * Created by Владимир on 16.08.2017.
 */
class SlidingWindow<T> @JvmOverloads constructor(size: Int, var offset: Int = 0) {
    private var values: Array<Any?>
    private var size = 0

    init {
        values = arrayOfNulls(size)
    }

    fun getSize(): Int {
        return values.size
    }

    operator fun set(number: Int, value: T) {
        val windowIndex = number - offset
        if (windowIndex > values.size - 1) {
            e("ARQ: window index out of bounds")
            return
        } else if (windowIndex < 0) return
        values[windowIndex] = value
    }

    fun advance(): T? {
        val firstBlock = values[0] as T? ?: return null
        System.arraycopy(values, 1, values, 0, values.size - 1)
        values[values.size - 1] = null
        offset++
        return firstBlock
    }

    fun getValue(windowIndex: Int): T? {
        return values[windowIndex] as T?
    }

    fun tail(): Int {
        return offset + values.size - 1
    }

    fun setSize(windowSize: Int) {
        val tempValues = arrayOfNulls<Any>(windowSize)
        System.arraycopy(values, 0, tempValues, 0, min(windowSize, size))
        values = tempValues
        size = windowSize
    }
}